src/backend/distributed/README.md

```diff
@@ -0,0 +1,2472 @@
+
+INSERT INTO users_table (user_id) VALUES (1) RETURNING (SELECT count(*) FROM users_table);
+ERROR:  subqueries are not supported within INSERT queries
+HINT:  Try rewriting your queries with 'INSERT INTO ... SELECT' syntax.
+```
+
+ **Simple Inserts with a Single VALUES Clause**:
+-- As noted in the "fast-path router planner", these INSERT commands are planned with fast-path planning. This does not require calling into `standard_planner()`, and the distribution key should be extracted from the query itself.
+```sql
+INSERT INTO users_table VALUES (1, 'onder', 'onderkalaci@gmail.com', now() - '5 years'::interval, 'TR');
+```
+
+ **Main Functions**:
+  The main functions involved in this path are `RegenerateTaskListForInsert()`, `FastPathRouterQuery()`, and `RouterInsertTaskList`. For single-row INSERT tasks, `Job->deferredPruning=true`, meaning we can always do the shard pruning during execution.
+
+ **Multi-row INSERTs**:
+For multi-row INSERTs, `RouterInsertTaskList()` becomes slightly more interesting. Citus groups rows by target shard.
+```sql
+INSERT INTO orders_table (order_id, user_id) VALUES
+	(1, 1), (2, 2), (3, 1), (4, 3), (5, 2);
+```
+
+ **Debug Info**:
+ Debug information shows how the query is rebuilt for different user_ids.
+```sql
+-- for user_id: 1
+DEBUG:  query after rebuilding:  INSERT INTO public.orders_table_102041 AS citus_table_alias (order_id, user_id) VALUES ('1'::bigint,'1'::bigint), ('3'::bigint,'1'::bigint)
+
+-- for user_id: 3
+DEBUG:  query after rebuilding:  INSERT INTO public.orders_table_102055 AS citus_table_alias (order_id, user_id) VALUES ('4'::bigint,'3'::bigint)
+
+-- for user_id: 2
+DEBUG:  query after rebuilding:  INSERT INTO public.orders_table_102064 AS citus_table_alias (order_id, user_id) VALUES ('2'::bigint,'2'::bigint), ('5'::bigint,'2'::bigint)
+```
+
+
+
+### INSERT.. SELECT and MERGE Command Query Planning
+
+ **Overview**:
+-- This section discusses `INSERT .. SELECT` and `MERGE` commands, which share almost identical planning logic.
+
+ **Planning Methods**:
+ Broadly, there are three methods to plan these commands:
+1. Pushdown
+2. Pull-to-coordinator
+3. Repartition
+
+ **Performance Considerations**:
+ When it comes to performance and resource utilization, pushdown is generally the most efficient. For handling large data sizes, the repartition method scales better than the pull-to-coordinator method.
+
+ **Further Reading**:
+ For more detailed information on pushdown and repartition methods, refer to this [blog post](https://www.citusdata.com/blog/2023/07/27/how-citus-12-supports-postgres-merge/). The post focuses on the `MERGE` command but is also applicable to `INSERT .. SELECT`.
+
+ **Examples**:
+ The following section will delve into examples, starting with simple ones and moving to more complex scenarios.
+
+
+Certainly, here's your text with formatting adjustments for improved readability. I've made sure not to alter any text, comments, or SQL code.
```

Comment:
Are these lines from chatgpt or something? I think they don't belong here, right?

Reply:
ah, good catch. I relied on chatgpt in many places while writing this